Give control to RaftedValue.Data and allow Raft Backed Databases

hansonkd opened the pull request:
This PR has several major breaking changes:

These changes are needed to use RaftedValue for values not contained in memory (ie a disk backed database)

1. Consenus data and value data are seperated in Snapshot
	* New Module SnapshotConsensus holds consensus information.
	* When Persisting
		* Consensus is saved under `[snapshot_dir]/consensus.snap`
		* Value is saved under `[snapshot_dir]/value.snap`
2. Allow values to be created in an environment.
    * `:data_environment` is now an option in `make_config/2`
    * `new/0` is now `new/1`
    * `:data_environment`'s value is passed to `new/1`
3. Do not make config a rafted value:
	* Making configuration a rafted value is potentially dangerous.
	   * LeaderHook and data_module could be changed to something hazardous.
	   * Config values should be version controlled
	* data_environment is unique to each node and cannot be shared.
	* removes `:change_config` message logic.
	* raft_fleet takes advantage of the rafted config to automatically setup nodes. 
		* This can be fixed by:
			1. Making the assumption that each fleet is responsible for 1 type of value.
			2. Changing `RaftFleet.activate("zone1")` to `RaftFleet.activate("zone1", config)`
			3. By making restrictions we are actually adding value. 
4. Allow Value module to control snapshot process
	* Callbacks when Persisting	
        * `from_disk/2`, `to_disk/1`
        * In a database you just want to copy the db file and gzip it.
        * The path passed into `from_disk/2` should be write protected. Modifying or moving this file will mess up the leader restoring from a local backup.
    * Callbacks when non-Persisting:
        * Adds `to_snapshot/1` `from_snapshot/2`
        * (These are mostly to keep parity with persistent callbacks)
    * `from_snapshot/2` and `from_disk/2` get passed `Config.data_environment`
5. Sending Compressed Snapshot Value Incrementally
    * New Module `Transfer`
    * `InstallCompressedSnapshot` now has `value_io_pid` which is a `pid` 
    * When a server is restoring from `InstallCompressedSnapshot` calls `Transfer.stream` on `pid` to stream remote file and save it in a temporary file.
        * Then reads this temporary file with `from_disk/2`


### LMDB Example
	

An example of building a disk backed KV store on rafted_value:


```ex
defmodule Database do
   defstruct [:dir, :db]

   def new([dir: dir]) do
   		# Open a db in the directory
   		db = open_db(dir, [:create])
   		%Database{db: db, dir: dir}
   end
   def command( %Database{db: db} = state, {:put, k, v}) do
   		{:elmdb.put(db, k, v), state}
   end
   def command( %Database{db: db} = state, {:get, k}) do
   		{:ok, v} = :elmdb.get(db, k)
   		{v, state}
   end
   def query(%Database{db: db} = state, {:get, k}) do
   		{:ok, v} = :elmdb.get(db, k)
   		v
   end
   
   # Only persist based snapshots allowed.
   def to_snapshot(data), do: raise "Not supported"
   def from_snapshot(ss, _), do: raise "Not supported"

   # Persisting to disk involves copying the  
   def to_disk(%Database{dir: dir}, path) do
   	  raw_db_location(dir)
   	  |> File.cp(path)
   end

   # Restoring from disk, copies the downloaded snapshot into the db directory
   def from_disk(path, [dir: dir]) do
   	  # Copy the snapshot to the data dir
   	  File.cp(path, raw_db_location(dir))
   	  db = open_db(dir)
   	  %Database{db: db, dir: dir}
   end

   # The file inside the db directory which stores the data
   defp raw_db_location(dir), do: Path.join(dir, 'data.mdb')
   defp open_db(dir, opts \\ []) do
   	    {:ok, env} = :elmdb.env_open(to_charlist(dir), [{:max_dbs, 3}])
        {:ok, db} = :elmdb.db_open(env, "db", opts)
        db
   end
end
```

hansonkd committed:
Give control to RaftedValue.Data

	1. Consenus data and value data are seperated in Snapshot
		* New Module SnapshotConsensus holds consensus information.
		* When Persisting
			* Consensus is saved under `[snapshot_dir]/consensus.snap`
			* Value is saved under `[snapshot_dir]/value.snap`
	2. Allow values to be created in an environment.
	    * `:data_environment` is now an option in `make_config/2`
	    * `new/0` is now `new/1`
	    * `:data_environment`'s value is passed to `new/1`
	3. Do not make config a rafted value:
		* Making configuration a rafted value is potentially dangerous.
		   * LeaderHook and data_module could be changed to something hazardous.
		   * Config values should be version controlled
		* data_environment is unique to each node and cannot be shared.
	4. Allow Value module to control snapshot process
		* Callbacks when Persisting
	        * `from_disk/2`, `to_disk/1`
	        * In a database you just want to copy the db file and gzip it.
	    * Callbacks when non-Persisting:
	        * Adds `to_snapshot/1` `from_snapshot/2`
	        * (These are mostly to keep parity with persistent callbacks)
	    * `from_snapshot/2` and `from_disk/2` get passed `Config.data_environment`
	5. Sending Compressed Snapshot Value Incrementally
	    * New Module `Transfer`
	    * `InstallCompressedSnapshot` now has `value_io_pid` which is a `pid`
	    * When a server is restoring from `InstallCompressedSnapshot` calls `Transfer.stream` on `pid` to stream remote file and save it in a temporary file.
	        * Then reads this temporary file with `from_disk/2`

hansonkd committed:
Fix path type annotations

hansonkd committed:
Remove :change_config message

skirino commented:
I appreciate your contribution. I think adding support of disk-backed state machine is intriguing.
Let me make some comments/questions:

- Inconsistent snapshots
    - Due to the original design that snapshot files are created by a separate `snapshot_writer` process, I think it's possible that, after making snapshot for the consensus logs and before making snapshot for the disk-resident data, the leader can accept subsequent commands and modify the disk-resident data, resulting in inconsistent snapshots.
    - Probably it's necessary that underlying storage engine supports something like "start making snapshot of current data without blocking subsequent writes"-functionality and each raft member process directly triggers snapshotting using this.
- About `change_config/2` and replicating `config`
    - Our team is maintaining a cluster of ErlangVMs that host approximately 50 consensus groups (using [raft_fleet](https://github.com/skirino/raft_fleet)). We definitely need a means to configure election/heartbeat timeout at runtime, so please don't remove it! They are important tuning parameters that must be adjusted according to load of each consensus group.
    - Regarding `config`, if I understand correctly, what you want to achieve is to add `data_environment` in each raft member process's internal state. For this purpose it's not necessary to include `data_environment` in `RaftedValue.Config` struct; instead it may be placed in e.g. `RaftedValue.Persistence` struct.
    - (I don't understand your comments on raft_fleet; please elaborate a bit more.)
- Some backward incompatible changes are unnecessarily introduced
    - For example, RPC structs are used by currently running raft member processes.
      Removing/renaming existing fields is backward incompatible (current version of rafted_value cannot understand the new format).

Please correct me if I misunderstand something!

hansonkd committed:
Make Config Rafted Again.

hansonkd committed:
Clean up for less diffs

hansonkd committed:
Add PersistenceHook to synchronize snapshot writes.

hansonkd committed:
Change log_compaction_rate back.

hansonkd committed:
Add documentation about `:persistence_hook`

hansonkd commented:
@skirino I thought about what you said and reverted the change to `Config`. I added `:data_environment` to `opts` next to `:persistence_dir` and it is rather nice since it is only used on server boot up, it doesn't need to be included in state.

I also add `:persistence_hook` since there are cases when you want to know that the snapshot is finished (right now used to coordinate the tests, possibly also to trigger an S3 upload)

For the inconsistent snapshots, you are right. There is a chance that snapshots can be inconsistent.


1. In the immediate term this can be solved by having the `Rafted.Data.command` log each command into its database also. Then when you restore from snapshot and replay, you discard all commands that are already in the db.
2. In the long term there is a possibility of making Log module a configuration so that logs could be stored in the same db as the data. SnapshotMeta information and data changes would be atomic. This would prevent duplicating the storage of entries. Thus when we take the snapshot, we are guaranteed that the logs match the data.
 

With regards to backwards compatibility there are two main issues: snapshot format and message format.


1. Snapshot format is easy. Just check if the snapshot dir is a folder or a file and process the single file snapshots normally. And save all snaphots in the new format.
2. `InstallSnapshotCompressed` is hard because of the new snapshot disk format I am not sure how to make `InstallSnapshotCompressed.bin` without reading the snapshot from disk and recompressing it. 
    * So in the transition `InstallSnapshotCompressed` will not be that efficient,
    * As an alternative we could do disregard the on_disk snapshot and simply run `%InstallSnapshotCompressed{bin: make_install_snapshot(state) |> Snapshot.encode()}` but again, that is slower and blocking.
    * We could write the snapshot in both formats, but again that is slower.
